### PR TITLE
3120: admin-panel-server entities route

### PR DESCRIPTION
**Pull request by chris-bes**
### Changes:

- Added entities route to admin-panel-server
- Allow filter in entity-server to have operators (eg: ==, !=,...)

### CODE REVIEW

File: packages/entity-server/src/routes/hierarchy/middleware/filter.ts
Comment (chris-bes) [inline packages/entity-server/src/routes/hierarchy/middleware/filter.ts:131]: This is a bit of a hack to merge the allowedCountries filter and the optional country_code filter (especially when we want to do `country_code` `!=`) without having to do any custom logic

File: packages/admin-panel-server/src/app/createApp.ts
Comment (chris-bes) [inline packages/admin-panel-server/src/app/createApp.ts:51]: We also need the root project entity code here to get all the entities of a hierarchy. I think the `/projects` endpoint in meditrak-server can supply that.

File: packages/admin-panel-server/src/routes/EntitiesRoute.ts
Comment (chris-bes) [inline packages/admin-panel-server/src/routes/EntitiesRoute.ts:22]: I'm thinking we should leave the `fields` query parameter for front end to decide. The reason is in case we want to implement ancestor chain, we will need to request for extra `child_codes` field?

File: packages/entity-server/src/models/types.ts
Comment (chris-bes) [inline packages/entity-server/src/models/types.ts:51]: Have to admit I'm not very confident with these types. I'm not also sure if they should sit in this file. I'm putting them in this file because they are closely related to the BaseModel

**Comment by chris-bes** [comment]
Will also update examples.http in entity-server once review is done

File: packages/entity-server/src/models/types.ts
Comment (rohan-bes) [reply packages/entity-server/src/models/types.ts:51]: You've actually done such an amazing job with these types! I know they're super hard to wrap your head around but I think you've got it almost perfect!

File: packages/entity-server/src/models/types.ts
Comment (rohan-bes) [inline packages/entity-server/src/models/types.ts:45]:
```suggestion
  [field in QueryConjunctions]?: FilterCriteria<T>;
```
I _believe_ that we support recursion in this config, eg:
```
{
  name: 'TO'
  _or_: {
    type: 'facility',
    _or_: {
      parent_id: '1234',
      location_type: 'area'
    }  
  }
}
=>
WHERE code = 'TO' OR (type = 'facility' OR (parent_id = '1234' and location_type = 'area'))
```

Lucky for us, type also supports all kind of recursive insanity 🎉

File: packages/entity-server/src/routes/hierarchy/middleware/filter.ts
Comment (rohan-bes) [inline packages/entity-server/src/routes/hierarchy/middleware/filter.ts:59]: Couple things about the types here:
1. `'==': '=', // Exact match` would also work right? I have a slight preference towards that because it reads a bit more clearly to me, and because typescript doesn't let you mess around with 'falsey' values as much as JS. ie. `null` != `undefined` 
2. Currently the type of this object is:
```
{
  '==': null;
  '!=': string;
  '=@': string;
};
```
The types of the keys are string literals (https://www.typescriptlang.org/docs/handbook/2/template-literal-types.html) whereas the values are just string (or null for the null one). This is how typescript works by default, but it does have it's drawbacks. 

For instance, later on you are trying to find the comparator for the operator by looking up in this table. Unfortunately the type will always be `string | null` at that point, which isn't really what you want. Instead you want the type to be `'!=' | 'ilike' | null` right?

Luckily there's a way to get this working in typescript as well!
```suggestion
const operatorToSqlComparator = {
  '==': '=' as const, // Exact match (switched to '=' here as it's my preference but if you stick with null theres no need for the 'as const')
  '!=': '!=' as const, // Does not match
  '=@': 'ilike' as const, // Contains sub string
};
```
This changes the type of the object to:
```
{
  '==': '=';
  '!=': '!=';
  '=@': 'ilike';
};
```
Which allows you to do typesafe lookups against this object!

File: packages/entity-server/src/routes/hierarchy/middleware/filter.ts
Comment (rohan-bes) [inline packages/entity-server/src/routes/hierarchy/middleware/filter.ts:60]:
```suggestion
type Operator = keyof typeof operatorToSqlComparator;

const filterOperators = Object.keys(operatorToSqlComparator) as Operator[];
```
`Object.keys()` is an area that typescript kinda drops the ball imo... It always returns type of `string[]` even tho it could theoretically return a string literal array like `('==' | '!=' | '=@')[]`. Oh well, best to cast it since we know what it'll be and then our types start playing nicely again 👍

File: packages/entity-server/src/routes/hierarchy/middleware/filter.ts
Comment (rohan-bes) [inline packages/entity-server/src/routes/hierarchy/middleware/filter.ts:62]:
```suggestion
const formatComparisonValue = (value: string | string[], operator: Operator) => {
```

File: packages/entity-server/src/routes/hierarchy/middleware/filter.ts
Comment (rohan-bes) [inline packages/entity-server/src/routes/hierarchy/middleware/filter.ts:70]:
```suggestion
const convertValueToAdvancedCriteria = (operator: Operator, value: string | string[]) => {
```
This one is important, as otherwise the lookup in the `operatorToSqlComparator` won't give us a typesafe response

File: packages/entity-server/src/routes/hierarchy/middleware/filter.ts
Comment (rohan-bes) [inline packages/entity-server/src/routes/hierarchy/middleware/filter.ts:80]: Once we cast `filterOperators` to `Operator[]` then the type of `operator` here becomes `Operator | undefined`. Doing the `if(!operator)` check later on then ensures that the type is `Operator` for the rest of the codeflow 👍

File: packages/entity-server/src/routes/hierarchy/middleware/filter.ts
Comment (rohan-bes) [inline packages/entity-server/src/routes/hierarchy/middleware/filter.ts:125]: When I checked out this branch, it was originally showing a pretty complex typescript error here. This is the downside of dealing with complex nested generic types as we are doing here, the compiler isn't great at explaining what went wrong... However I think the real issue is that we weren't being typesafe with the `comparator` field, which the `EntityFilter` type expected to have the explicit `comparator: '!=' | 'ilike'`, whereas we were giving it `comparator: string | null`. I think if you follow all the suggestions I made here, then it should clear up the error (atleast it has in my local). Let me know if it doesn't and we can work on it together 👍

File: packages/entity-server/src/models/types.ts
Comment (rohan-bes) [inline packages/entity-server/src/models/types.ts:9]:
```suggestion
type FilterComparators = '!=' | 'ilike' | '=';
```
See comments about this in `filter.ts`

**Review by rohan-bes — CHANGES_REQUESTED** [review]
Looks amazing @chris-bes, nice work!

Totally happy with the functionality, just suggestions on how to resolve the ts errors and ensure the types match the model.

So nice to see someone else getting stuck into writing some pretty complex types, I think you did a great job!

**Comment by rohan-bes** [comment]
Oh sorry @chris-bes one last request! Could you possibly break this PR into?
- Changes to entity-server/aggregator
- Changes to admin-panel-server

I'm totally cool with the changes to admin-panel-server staying on the epic branch, but I'd prefer if the entity-server/aggreator changes go straight to dev? There's a few refactors going on around entity-server right now and I'd prefer to incorporate your changes sooner rather than later!

**Comment by chris-bes** [comment]
@rohan-bes yep good idea! Thank you for your throughout review as well

File: packages/entity-server/src/models/types.ts
Comment (chris-bes) [reply packages/entity-server/src/models/types.ts:34]: I've actually tried that before. However typescript couldn't resolve the keys coming from @tupaia/database. So that's why I sorta gave up and just created an enum here :(:

image

File: packages/entity-server/src/routes/hierarchy/middleware/filter.ts
Comment (chris-bes) [reply packages/entity-server/src/routes/hierarchy/middleware/filter.ts]: The reason I had `null` was to avoid having the advanced criteria if the operator is `==`. So I still prefer the filter object to be:
```
{
     type: 'district'
}
```
instead of always
```
{
     type: {
           comparator: '=',
           comparisonValue: 'district'
     }
}
```

However, I've just realized I didn't add the logic to follow the 1st format mentioned above (I was working on this quite late so that might explain 😄 ). I've followed your `=` suggestion, but still have an if check for `=` operator to avoid using comparator/comparisonValue if possible

File: packages/entity-server/src/routes/hierarchy/middleware/filter.ts
Comment (chris-bes) [reply packages/entity-server/src/routes/hierarchy/middleware/filter.ts:137]: I think that was the problem! Thanks heaps!

File: packages/entity-server/src/routes/hierarchy/middleware/filter.ts
Comment (chris-bes) [reply packages/entity-server/src/routes/hierarchy/middleware/filter.ts:143]: I've checked and I think it still works, but in a coincident way(?), so I tweaked the code in ResponseBuilder.ts to fit with this. Hopefully that's alright

File: packages/entity-server/src/routes/hierarchy/middleware/filter.ts
Comment (chris-bes) [inline packages/entity-server/src/routes/hierarchy/middleware/filter.ts:125]: What do you think about return this as default if there's no filter instead of the previous:
```
{ country_code: allowedCountries }
```

When it was returning the previous format, I found it hard to debug the country_code filter because I kinda got confused a lot regarding whether country_code is something from the query filter or is it the default allowed countries. So I just change it to return this format (consistent with how we have the default filter when merging at the end of this function). So when I see this, then I know it's the default allowed countries filter.

**Comment by chris-bes** [comment]
Have broken up this PR into 2 commits:
https://github.com/beyondessential/tupaia/pull/2860/commits/986f137ccc0a99229d8f6331209f9ddd23c7e16c: entity-server change
https://github.com/beyondessential/tupaia/pull/2860/commits/986f137ccc0a99229d8f6331209f9ddd23c7e16c: admin-panel-server change.

I accidently did a force push instead of creating a new PR so sorry if it makes re-review harder. We can cherry-pick the first commit to a separate PR. Then when it passes testing and gets merged to dev, I can merge dev back and resolve the merge conflicts if any.

File: packages/entity-server/src/models/types.ts
Comment (rohan-bes) [reply packages/entity-server/src/models/types.ts:34]: Ah! Perhaps try `[field in keyof typeof QUERY_CONJUNCTIONS]`? I think I mistyped

**Review by rohan-bes — APPROVED** [review]
All looks good @chris-bes, excellent work!

My only standing request is that the changes to entity-server get merged straight to dev? (happy to give the PR a review when you break it up). Unless viz-builder-epic-phase-1 is gonna get merged to dev real soon, then I guess I don't mind as much 👍

**Comment by chris-bes** [comment]
> My only standing request is that the changes to entity-server get merged straight to dev? (happy to give the PR a review when you break it up). Unless viz-builder-epic-phase-1 is gonna get merged to dev real soon, then I guess I don't mind as much 👍

Yep I was waiting for approval to cherry-pick into another PR. I think it will also require some regression testing, so happy to break it into another task

**Pull request merged**
